libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,149 @@
+	for (auto&& [index, item]: _items | ranges::views::enumerate)
+	{
+		// The number of PushTags approximates the number of calls to a block.
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
+		// Using breaksCSEAnalysisBlock will hopefully allow the return jump to be optimized after inlining.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, false))
+		{
+			if (item == Instruction::JUMP)
+				inlinableBlockItems[_items[*lastTag].data()] = _items | ranges::views::slice(*lastTag + 1, index + 1);
+			lastTag.reset();
+		}
+
+		if (item.type() == Tag)
+			lastTag = index;
+	}
+
+	// Filter candidates for general inlinability and store the number of PushTags alongside the assembly items.
+	map<u256, InlinableBlock> result;
+	for (auto&& [tag, items]: inlinableBlockItems)
+		if (uint64_t const* numPushes = util::valueOrNullptr(numPushTags, tag))
+		{
+			InlinableBlock block{items, *numPushes};
+			if (isInlineCandidate(tag, block))
+				result.emplace(std::make_pair(tag, block));
+		}
+	return result;
+}
+
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
+{
+	AssemblyItem exitJump = _block.items.back();
+
+	if (
+		_jump.getJumpType() == AssemblyItem::JumpType::IntoFunction &&
+		exitJump.getJumpType() == AssemblyItem::JumpType::OutOfFunction
+	)
+	{
+		exitJump.setJumpType(AssemblyItem::JumpType::Ordinary);
+		size_t codeSize = static_cast<size_t>(_block.items.size()) - 1;
+		bigint uninlinedCost = bigint(m_runs) * 24u * _block.pushTagCount + (8u * _block.pushTagCount + 2 + codeSize) * 200u;
+		bigint inlinedCost = _block.pushTagCount * codeSize * 200;
+
+		if (uninlinedCost > inlinedCost)
+			return exitJump;
+	}
+
+	return nullopt;
+}
+
+
+void Inliner::optimise()
+{
+	std::map<u256, InlinableBlock> inlinableBlocks = determineInlinableBlocks(m_items);
+
+	if (inlinableBlocks.empty())
+		return;
+
+	AssemblyItems newItems;
+	for (auto it = m_items.begin(); it != m_items.end(); ++it)
+	{
+		AssemblyItem const& item = *it;
+		if (next(it) != m_items.end())
+		{
+			AssemblyItem const& nextItem = *next(it);
+			if (item.type() == PushTag && nextItem == Instruction::JUMP)
+				if (auto *inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
+					if (auto exitJump = shouldInline(item.data(), nextItem, *inlinableBlock))
+					{
+						newItems += ranges::views::drop_last(inlinableBlock->items, 1);
+						newItems.emplace_back(*exitJump);
+
+						// We are removing one push tag to the block we inline.
+						--inlinableBlock->pushTagCount;
+						// We might increase the number of push tags to other blocks.
+						for (AssemblyItem const& inlinedItem: inlinableBlock->items)
+							if (inlinedItem.type() == PushTag)
+								if (auto* block = util::valueOrNullptr(inlinableBlocks, inlinedItem.data()))
+									++block->pushTagCount;
+
```

Comment:
I don't really follow how `pushTagCount` is used. Can you please explain?

Reply:
``pushTagCount`` counts how many times the tag at which the respective inline candidate lives is pushed, i.e. roughly how many jumps there are into that block. When inlining we can increase the number of pushes, though, that's accounted for here.
Unlikely to happen currently, since we restrict to full function inlining, but more likely to happen in the more general case and still not impossible in full function inlining.